### PR TITLE
Introduce debug-process-names and experimental-enable-dwarf-unwinding

### PR DESCRIPTION
**Pull request by kakkoyun**
Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

```
// Hidden debug flags (only for debugging):
--debug-process-names []string 
Only attach profilers to specified processes. comm name will be used to match the given matchers. Accepts Go regex syntax (https://pkg.go.dev/regexp/syntax).

// These flags are experimental. Use them at your own peril.
--experimental-enable-dwarf-unwinding bool 
Unwind stack using .eh_frame information.
```
## Test

1. `sudo ./scripts/local-run.sh` with **`testdata/out/basic-cpp-no-fp`**:

<details>

```shell
sudo "${PARCA_AGENT}" \
            --node=local-test \
            --http-address=:7071 \
            --log-level=debug \
            --experimental-enable-dwarf-unwinding \
            --debug-process-names=basic-cpp-no-fp \
            --remote-store-address=localhost:7070 \
            --remote-store-insecure
```
CleanShot 2022-11-17 at 10 54 19@2x

</details>

2.  `sudo ./scripts/local-run.sh` with **CRuby interpreter**:

<details>

```shell
sudo "${PARCA_AGENT}" \
            --node=local-test \
            --http-address=:7071 \
            --log-level=debug \
            --experimental-enable-dwarf-unwinding \
            --debug-process-names=ruby \
            --remote-store-address=localhost:7070 \
            --remote-store-insecure
```

CleanShot 2022-11-17 at 14 59 16@2x

CleanShot 2022-11-17 at 14 59 02@2x

</details>

3. `make dev/up` with Minikube

<details>

CleanShot 2022-11-16 at 16 48 16

CleanShot 2022-11-16 at 16 51 04@2x

</details>

### CODE REVIEW

File: pkg/profiler/cpu/cpu.go [hunk 1]
```diff
@@ -263,54 +271,40 @@ func (p *CPU) Run(ctx context.Context) error {
...
+			return fmt.Errorf("failed to create procfs: %w", err)
+		}
 
-	dwarfStackTraces, err := m.GetMap(dwarfStackTracesMapName)
-	if err != nil {
-		return fmt.Errorf("get dwarf stack traces map: %w", err)
-	}
+		var matchers []*regexp.Regexp
+		for _, exp := range p.debugProcessNames {
+			regex, err := regexp.Compile(exp)
```
Comment (brancz) [inline pkg/profiler/cpu/cpu.go:288]: compiling a regex is rather expensive, we should do this once only

File: cmd/parca-agent/main.go [hunk 2]
```diff
@@ -108,8 +108,11 @@ type flags struct {
 	DebuginfoStrip               bool          `kong:"help='Only upload information needed for symbolization. If false the exact binary the agent sees will be uploaded unmodified.',default='true'"`
 	DebuginfoUploadCacheDuration time.Duration `kong:"help='The duration to cache debuginfo upload exists checks for.',default='5m'"`
 
+	// Hidden debug flags (only for debugging):
+	DebugProcessNames []string `kong:"help='Only attach profilers to specified processes. comm name will be used to match the given matchers. Accepts Go regex syntax (https://pkg.go.dev/regexp/syntax).',hidden=''"`
```
Comment (brancz) [inline cmd/parca-agent/main.go:112]: Do we really need this to be a list? A single regex would be sufficient, no? If people really want to they can just use "or" in the regex.

**Review by brancz — COMMENTED** [review]
I personally think a set of `comm`s would have been sufficient, but I'm happy with the regex as well. Only the compiling comment is something I think we need to address the rest I don't feel strongly about.

**Comment by brancz** [comment]
Since it's all experimental and we can change flags arbitrarily especially with experimental features, let's just make this a single flag I would say.

File: pkg/profiler/cpu/cpu.go (hunk 1 above)
Comment (kakkoyun) [reply pkg/profiler/cpu/cpu.go:288]: This actually happens once. It's outside the profiling loop and the watch loop.

File: cmd/parca-agent/main.go (hunk 2 above)
Comment (kakkoyun) [reply cmd/parca-agent/main.go:112]: This is true. I also wanted to give the option to match against the exact names without thinking of the regex syntax. e.g. `etcd, kube-apiserver`

**Comment by kakkoyun** [comment]
> Since it's all experimental and we can change flags arbitrarily especially with experimental features, let's just make this a single flag I would say.

One of the reasons, I want to have 2 flags, is to be able to use `debug-process-names` solely. It has many use cases while debugging things with the agent. Especially, after enabling system-wide profiling the information we collect from even a single node system can be overwhelming while trying to debug some specific thing.

File: bpf/cpu/cpu.bpf.c [hunk 3]
```diff
@@ -162,6 +162,8 @@ u32 UNWIND_SAMPLES_COUNT = 7;
 
 /*================================ MAPS =====================================*/
 
+BPF_HASH(configuration, u8, u8, 1);
```
Comment (kakkoyun) [inline bpf/cpu/cpu.bpf.c:165]: @javierhonduco Do you know a more space-efficient way for these types of maps? I don't need the values; I want to check the existence of the keys.

**Review by brancz — APPROVED** [review]

**Comment by javierhonduco** [comment]
> I personally think a set of comms would have been sufficient, but I'm happy with the regex as well. Only the compiling comment is something I think we need to address the rest I don't feel strongly about.

Having a regex is very useful for some applications that use named threads, they typically name them as `io-thread-$IO_THREAD_COUNT` :D

**Comment by javierhonduco** [comment]
Oh, the only other thing I wanted to ask is to test if everything works for dwarf based stack unwinding. Could you run `cpp-basic-no-fp` and some simple Ruby script to check that everything's working fine? 

The output should not be too different from https://github.com/parca-dev/parca-agent/pull/978's test plan

**Comment by kakkoyun** [comment]
> Oh, the only other thing I wanted to ask is to test if everything works for dwarf based stack unwinding. Could you run `cpp-basic-no-fp` and some simple Ruby script to check that everything's working fine?
> 
> The output should not be too different from #978 test plan

Updated the issue with the test results.

**Review by javierhonduco — APPROVED** [review]
LGTM!! 🎉

**Pull request merged**